File: 53-MaximumSubarray/53-MaximumSubarray.java
// Last updated: 14/08/2025, 11:34:49
public class Solution {
    public int maxSubArray(int[] nums) {
        // Initialize variables
        int maxCurrent = nums[0]; // The current subarray sum
        int maxGlobal = nums[0];  // The maximum sum found so far
        
        // Iterate through the array starting from the second element
        for (int i = 1; i < nums.length; i++) {
            // Update the current subarray sum
            maxCurrent = Math.max(nums[i], maxCurrent + nums[i]);
            // Update the global maximum sum
            if (maxCurrent > maxGlobal) {
                maxGlobal = maxCurrent;
            }
        }
        
        return maxGlobal; // Return the maximum subarray sum
    }
}

File: 53-MaximumSubarray/53-MaximumSubarray.py
# Last updated: 09/09/2026, 16:36:30
class Solution:
    def maxSubArray(self, nums: List[int]) -> int:
        currsum=nums[0]
        maxsum=nums[0]

        for num in nums[1:]:
            currsum=max(num,currsum+num)
            maxsum=max(maxsum,currsum)
        return maxsum    